`change` event for input[type=file] does not fire when different file with same name is selected

rr-codes opened the pull request:
#### 14b28281c8d06ca7fed8f17777c1f071d4e52d5f
<pre>
`change` event for input[type=file] does not fire when different file with same name is selected
<a href="https://bugs.webkit.org/show_bug.cgi?id=187084">https://bugs.webkit.org/show_bug.cgi?id=187084</a>
rdar://98460303

Reviewed by Ryosuke Niwa and Aditya Keerthi.

When a different file with the same name as the currently selected file is selected,
the `change` event does not currently fire, because a file&apos;s path is it&apos;s
identifier, so the two files are considered the same by the system.

This PR works around this by saving each file&apos;s unique id and then comparing the files
both based on their name and now their id too.

* Source/WTF/wtf/FileSystem.h:
* Source/WTF/wtf/glib/FileSystemGlib.cpp:
(WTF::FileSystemImpl::fileID):
(WTF::FileSystemImpl::fileIDsAreEqual):
(WTF::FileSystemImpl::08c7e637e60a):
(WTF::FileSystemImpl::fileCreationTime): Deleted.
* Source/WTF/wtf/posix/FileSystemPOSIX.cpp:
(WTF::FileSystemImpl::fileID):
(WTF::FileSystemImpl::fileIDsAreEqual):
* Source/WTF/wtf/win/FileSystemWin.cpp:
(WTF::FileSystemImpl::fileID):
(WTF::FileSystemImpl::fileIDsAreEqual):
* Source/WebCore/fileapi/File.cpp:
(WebCore::File::create):
(WebCore::File::File):
* Source/WebCore/fileapi/File.h:
* Source/WebCore/html/FileInputType.cpp:
(WebCore::FileInputType::setFiles):
(WebCore::FileInputType::filesChosen):
* Source/WebCore/platform/FileChooser.cpp:
(WebCore::FileChooser::chooseFiles):
(WebCore::FileChooser::chooseMediaFiles):

Canonical link: <a href="https://commits.webkit.org/255297@main">https://commits.webkit.org/255297@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c01ab766ff6eb3eb8ccc07b22ebf497368c9bd95

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/91924 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/1168 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/22462 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/101583 "Built successfully") | [❌ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/161770 "Reverted pull request changes (failure)") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/95926 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/1164 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/29622 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/84235 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/97966 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/97582 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/708 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/78475 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/27671 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/82630 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/82233 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/70706 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/83291 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/36017 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/16252 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/78342 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/33753 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/17352 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/27038 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3669 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/37630 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/40067 "Found 3 new test failures: imported/w3c/web-platform-tests/import-maps/multiple-import-maps/basic.html, imported/w3c/web-platform-tests/webcodecs/videoDecoder-codec-specific.https.any.html?h264_annexb, imported/w3c/web-platform-tests/webcodecs/videoDecoder-codec-specific.https.any.worker.html?h264_annexb (failure)") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/80963 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/39514 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/36471 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/17759 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WTF/wtf/glib/FileSystemGlib.cpp:
```diff
@@ -134,6 +134,18 @@ std::optional<uint64_t> fileSize(PlatformFileHandle handle)
+    return std::nullopt;
+}
+
+bool fileIDsAreEqual(std::optional<PlatformFileID> a, std::optional<PlatformFileID> b)
```
Is this method needed? Wouldn't `std::nullopt == std::nullopt`, for platforms that don't implement `fileID`?

rr-codes replied:
For POSIX and GLIB, `PlatformFileID` is a scalar, and so `operator ==` for `std::optional< PlatformFileID>` compiles fan because it transitively compares the two scalar values. But for Windows, this is not the case, since `PlatformFileID` is a struct, and so it fails to compile.

rniwa replied:
Why don't we use operator== instead? It seems wasteful that this is a full function call in platforms where it's a 64-bit scalar value. We might want to just in-line it in platforms other than Windows.

rr-codes replied:
> We might want to just in-line it in platforms other than Windows.

@rniwa  How would I do this?

aperezdc replied:
FTR, I have the GLib implementation of `FileSystem::fileID()` in https://github.com/WebKit/WebKit/pull/5081 and I'm using `CString` there as the type of the identifier.

rr-codes replied:
@aperezdc ah ok; does Cstring support `operator ==`?

cdumez replied:
> @aperezdc ah ok; does Cstring support `operator ==`?

Yes.

pxlcoder commented on Source/WTF/wtf/glib/FileSystemGlib.cpp:
```diff
@@ -134,6 +134,18 @@ std::optional<uint64_t> fileSize(PlatformFileHandle handle)
+
+bool fileIDsAreEqual(std::optional<PlatformFileID> a, std::optional<PlatformFileID> b)
+{
+    // FIXME (246118): Implement this function properly.z
```
Typo.

rr-codes committed:
`change` event for input[type=file] does not fire when different file with same name is selected
https://bugs.webkit.org/show_bug.cgi?id=187084
rdar://98460303

Reviewed by Ryosuke Niwa and Aditya Keerthi.

When a different file with the same name as the currently selected file is selected,
the `change` event does not currently fire, because a file's path is it's
identifier, so the two files are considered the same by the system.

This PR works around this by saving each file's unique id and then comparing the files
both based on their name and now their id too.

* Source/WTF/wtf/FileSystem.h:
* Source/WTF/wtf/glib/FileSystemGlib.cpp:
(WTF::FileSystemImpl::fileID):
(WTF::FileSystemImpl::fileIDsAreEqual):
(WTF::FileSystemImpl::08c7e637e60a):
(WTF::FileSystemImpl::fileCreationTime): Deleted.
* Source/WTF/wtf/posix/FileSystemPOSIX.cpp:
(WTF::FileSystemImpl::fileID):
(WTF::FileSystemImpl::fileIDsAreEqual):
* Source/WTF/wtf/win/FileSystemWin.cpp:
(WTF::FileSystemImpl::fileID):
(WTF::FileSystemImpl::fileIDsAreEqual):
* Source/WebCore/fileapi/File.cpp:
(WebCore::File::create):
(WebCore::File::File):
* Source/WebCore/fileapi/File.h:
* Source/WebCore/html/FileInputType.cpp:
(WebCore::FileInputType::setFiles):
(WebCore::FileInputType::filesChosen):
* Source/WebCore/platform/FileChooser.cpp:
(WebCore::FileChooser::chooseFiles):
(WebCore::FileChooser::chooseMediaFiles):

Canonical link: https://commits.webkit.org/255297@main
